openedx/core/djangoapps/content/learning_sequences/api/outlines.py

```diff
@@ -0,0 +1,323 @@
+from datetime import datetime
+from typing import Optional
+
+import attr
+from django.contrib.auth import get_user_model
+from django.db import transaction
+from edx_django_utils.cache import TieredCache, get_cache_key
+from edx_django_utils.monitoring import function_trace
+from opaque_keys.edx.keys import CourseKey, UsageKey
+
+from .data import (
+    CourseOutlineData, CourseSectionData, CourseLearningSequenceData,
+    UserCourseOutlineData, UserCourseOutlineDetailsData, VisibilityData,
+)
+from ..models import (
+    CourseSection, CourseSectionSequence, LearningContext, LearningSequence
+)
+from .permissions import can_see_all_content
+from .processors.schedule import ScheduleOutlineProcessor
+from .processors.visibility import VisibilityOutlineProcessor
+
+User = get_user_model()
+log = logging.getLogger(__name__)
+
+# Public API...
+__all__ = [
+    'get_course_outline',
+    'get_user_course_outline',
+    'get_user_course_outline_details',
+    'replace_course_outline',
+]
+
+
+def get_course_outline(course_key: CourseKey) -> CourseOutlineData:
+    """
+    Get the outline of a course run.
+
+    There is no user-specific data or permissions applied in this function.
+
+    See the definition of CourseOutlineData for details about the data returned.
+    """
+    learning_context = _get_learning_context_for_outline(course_key)
+
+    # Check to see if it's in the cache.
+    cache_key = "learning_sequences.api.get_course_outline.v1.{}.{}".format(
+        learning_context.context_key, learning_context.published_version
+    )
+    outline_cache_result = TieredCache.get_cached_response(cache_key)
+    if outline_cache_result.is_found:
+        return outline_cache_result.value
+
+    # Fetch model data, and remember that empty Sections should still be
+    # represented (so query CourseSection explicitly instead of relying only on
+    # select_related from CourseSectionSequence).
+    section_models = CourseSection.objects \
+        .filter(learning_context=learning_context) \
+        .order_by('ordering')
+    section_sequence_models = CourseSectionSequence.objects \
+        .filter(learning_context=learning_context) \
+        .order_by('ordering') \
+        .select_related('sequence')
+
+    # Build mapping of section.id keys to sequence lists.
+    sec_ids_to_sequence_list = defaultdict(list)
+
+    for sec_seq_model in section_sequence_models:
+        sequence_model = sec_seq_model.sequence
+        sequence_data = CourseLearningSequenceData(
+            usage_key=sequence_model.usage_key,
+            title=sequence_model.title,
+            visibility=VisibilityData(
+                hide_from_toc=sec_seq_model.hide_from_toc,
+                visible_to_staff_only=sec_seq_model.visible_to_staff_only,
+            )
+        )
+        sec_ids_to_sequence_list[sec_seq_model.section_id].append(sequence_data)
+
+    sections_data = []
+    for section_model in section_models:
+        sections_data.append(
```

Comment:
Nit: could be a list comprehension.

Reply:
Ah, right. Yeah, I think this was an artifact from before I moved the visibility information to the leaf nodes (I previously had separate sets for visibility/hide_from_toc which complicated this). Will update.